distsql/select_result.go

```diff
@@ -297,3 +318,95 @@ func (r *selectResult) Close() error {
 	}
 	return r.resp.Close()
 }
+
+type HasCopRuntimeStats interface {
+	// GetCopRuntimeStats gets the cop runtime stats information.
+	GetCopRuntimeStats() *tikv.CopRuntimeStats
+}
+
+type selectResultRuntimeStats struct {
+	execdetails.RuntimeStats
+	copRespTime      []time.Duration
+	procKeys         []int64
+	backoffSleep     map[string]time.Duration
+	backoffTimes     map[string]int
+	totalProcessTime time.Duration
+	totalWaitTime    time.Duration
+	rpcStat          tikv.RegionRequestRuntimeStats
+}
+
+func (s *selectResultRuntimeStats) mergeCopRuntimeStats(copStats *tikv.CopRuntimeStats, respTime time.Duration) {
+	s.copRespTime = append(s.copRespTime, respTime)
+	s.procKeys = append(s.procKeys, copStats.ProcessedKeys)
+
+	for k, v := range copStats.BackoffSleep {
+		s.backoffSleep[k] += v
+	}
+	for k, v := range copStats.BackoffTimes {
+		s.backoffTimes[k] += v
+	}
+	s.totalProcessTime += copStats.ProcessTime
+	s.totalWaitTime += copStats.WaitTime
+	s.rpcStat.Merge(copStats.RegionRequestRuntimeStats)
+}
+
+func (s *selectResultRuntimeStats) String() string {
+	buf := bytes.NewBuffer(nil)
+	if s.RuntimeStats != nil {
+		buf.WriteString(s.RuntimeStats.String())
+	}
+	if len(s.copRespTime) > 0 {
+		size := len(s.copRespTime)
+		buf.WriteString(", ")
+		if size == 1 {
+			buf.WriteString(fmt.Sprintf("cop_task: {num: 1, max:%v} proc_keys: {max: %v}", s.copRespTime[0], s.procKeys[0]))
+		} else {
+			sort.Slice(s.copRespTime, func(i, j int) bool {
+				return s.copRespTime[i] < s.copRespTime[j]
+			})
+			vMax, vMin := s.copRespTime[size-1], s.copRespTime[0]
+			vP80, vP95 := s.copRespTime[size*4/5], s.copRespTime[size*19/20]
+			sum := 0.0
+			for _, t := range s.copRespTime {
+				sum += float64(t)
+			}
+			vAvg := time.Duration(sum / float64(size))
+
+			sort.Slice(s.procKeys, func(i, j int) bool {
+				return s.procKeys[i] < s.procKeys[j]
+			})
+			keyMax := s.procKeys[size-1]
+			keyP95 := s.procKeys[size*19/20]
+			buf.WriteString(fmt.Sprintf("cop_task: {num: %v, max :%v, min:%v, avg:%v, p80:%v, p95:%v}, proc_keys: {max:%v, p95:%v}", size, vMax, vMin, vAvg, vP80, vP95, keyMax, keyP95))
+		}
+	}
+
+	rpcStatsStr := s.rpcStat.String()
+	if len(rpcStatsStr) > 0 {
+		buf.WriteString(", ")
+		buf.WriteString(rpcStatsStr)
+	}
+
+	if s.totalProcessTime > 0 {
+		buf.WriteString(", ")
+		buf.WriteString(fmt.Sprintf("Coprocessor: {total_process_time: %s", s.totalProcessTime.String()))
+		if s.totalWaitTime > 0 {
+			buf.WriteString(", ")
+			buf.WriteString("total_wait_time: ")
+			buf.WriteString(s.totalWaitTime.String())
+		}
+		buf.WriteString("}")
+	}
+
+	if len(s.backoffSleep) > 0 {
+		buf.WriteString(", ")
+		for k, v := range s.backoffTimes {
+			if buf.Len() > 0 {
+				buf.WriteByte(',')
+			}
+			d := s.backoffSleep[k]
+			buf.WriteString(fmt.Sprintf("%s_backoff:{num:%d, total_time:%s}", k, v, d.String()))
```

Comment:
How about combining them into one field: `backoff: {type1_num: ..., type1_time: ..., type2_num: ..., type2_time: ...}`?

Reply:
Em... I still prefer the current display. @wjhuang2016 PTAL